### PR TITLE
FailedLeader does not add follower - BTS-1162

### PR DESCRIPTION
**Pull request by neunhoef**
### Scope & Purpose

We found that a FailedLeader job automatically deploys a new follower.
This is probably not desirable. An AddFollower will do the work later
anyway. Therefore this proposes to make the new follower configurable.

- [*] :hankey: Bugfix
- [*] :pizza: New feature

### Checklist

- [*] Tests
  - [*] C++ **Unit tests**
- [*] :book: CHANGELOG entry made
- [*] Backports
   - [*] Backport for 3.10: https://github.com/arangodb/arangodb/pull/17798
   - [*] Backport for 3.9: https://github.com/arangodb/arangodb/pull/17799
   - [*] Backport for 3.9.6 branch (for 3.9.6.2): https://github.com/arangodb/arangodb/pull/17800

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [*] GitHub issue / Jira ticket: https://arangodb.atlassian.net/browse/BTS-1162
- [*] Design document: https://github.com/arangodb/documents/pull/123

### CODE REVIEW

**Review by maierlars — APPROVED** [review]

File: CHANGELOG [hunk 1]
```diff
@@ -1,6 +1,14 @@
 devel
 -----
 
+* Added a configuration option (for the agency):
+    --agency.supervision-failed-leader-adds-follower
+  with a default of `true` (behaviour as before). If set to `false`,
```
Comment (jsteemann) [inline CHANGELOG:6]: BE => AE
```suggestion
  with a default of `true` (behavior as before). If set to `false`,
```

File: arangod/Agency/Agent.cpp [hunk 2]
```diff
@@ -2530,6 +2530,14 @@ void Agent::updateSomeConfigValues(velocypack::Slice data) {
     _config.setSupervisionDelayFailedFollower(u);
     _supervision->setDelayFailedFollower(u);
   }
+  slice = data.get("failedLeaderAddsFollower");
+  if (slice.isBool()) {
+    bool b = slice.getBool();
+    LOG_TOPIC("12345", DEBUG, Logger::SUPERVISION)
+        << "Updating failedLeaderAddsFollower to " << b;
+    _config.setSupervisionDelayFailedFollower(b);
```
Comment (jsteemann) [inline arangod/Agency/Agent.cpp:2538]: Shouldn't this be:
```suggestion
    _config.setSupervisionFailedLeaderAddsFollower(b);
```

Comment (maierlars) [reply arangod/Agency/Agent.cpp:2538]: I hate C++. Why does that even compile ... yeah because you can implicitly convert ints to bools. :headdesk:

Comment (maierlars) [reply arangod/Agency/Agent.cpp:2538]: good catch btw

File: arangod/Agency/AgentConfiguration.h [hunk 3]
```diff
@@ -252,6 +257,11 @@ struct config_t {
     _supervisionDelayFailedFollower = d;
   }
 
+  /// @brief set Supervision FailedLeader adds follower flag
+  void setSupervisionFailedLeaderAddsFollower(bool f) {
+    _supervisionFailedLeaderAddsFollower = f;
```
Comment (jsteemann) [inline arangod/Agency/AgentConfiguration.h:262]: It is a bit shady that reading attributes from the configuration object is done under a read-lock and writes to the configuration object are done under to write-lock, except for these setters...
I think to be consistent, all these setters should acquire the write-lock as well.

File: CHANGELOG (hunk 1 above)
Comment (neunhoef) [reply CHANGELOG:6]: Will do.

File: arangod/Agency/Agent.cpp (hunk 2 above)
Comment (neunhoef) [reply arangod/Agency/Agent.cpp:2538]: Oh dear. Done.

File: arangod/Agency/AgentConfiguration.h (hunk 3 above)
Comment (neunhoef) [reply arangod/Agency/AgentConfiguration.h:262]: Agreed.

File: tests/Agency/FailedLeaderTest.cpp [hunk 4]
```diff
@@ -1139,6 +1197,200 @@ TEST_F(FailedLeaderTest, job_should_be_written_to_pending) {
...
+        result->add(VPackValue((uint64_t)1));
+        result->close();
+        return trans_ret_t(true, "1", 1, 0, result);
+      });
+  When(Method(mockAgent, waitFor))
+      .AlwaysReturn(AgentInterface::raft_commit_t::OK);
+  AgentInterface& agent = mockAgent.get();
+
+  // new server will randomly be selected...so seed the random number generator
+  srand(1);
```
[the remark this answers is not in the document]
Comment (neunhoef) [reply tests/Agency/FailedLeaderTest.cpp]: Done in all places.

**Review by jsteemann — APPROVED** [review]
LGTM

**Pull request merged**